{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CFDE C2M2 Model Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bento_meta.mdb.mdb_tools import ToolsMDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DSS Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import clean DSS mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mapping_df = pd.read_csv(\"Mappings/CRDC DST Cross-Model Mapping_clean.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(303, 6)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mapping_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>['case']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>GDC</td>\n",
       "      <td>id</td>\n",
       "      <td>['case']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>['case']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>PDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>['case']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ent_1_model ent_1_handle ent_1_extra_handles ent_2_model ent_2_handle  \\\n",
       "0         DST   Identifier         ['Subject']        CTDC      case_id   \n",
       "1         DST   Identifier         ['Subject']         GDC           id   \n",
       "2         DST   Identifier         ['Subject']        ICDC      case_id   \n",
       "3         DST   Identifier         ['Subject']         IDC   Patient ID   \n",
       "4         DST   Identifier         ['Subject']         PDC      case_id   \n",
       "\n",
       "  ent_2_extra_handles  \n",
       "0            ['case']  \n",
       "1            ['case']  \n",
       "2            ['case']  \n",
       "3            ['file']  \n",
       "4            ['case']  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mapping_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Filter for only C2M2 model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "c2m2_mapping_df = dss_mapping_df[dss_mapping_df[\"ent_2_model\"] == \"C2M2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "c2m2_mapping_df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19, 6)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c2m2_mapping_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>['Specimen']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>assay_type</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DST</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>['subject']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DST</td>\n",
       "      <td>Race</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>race</td>\n",
       "      <td>['subject_race']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DST</td>\n",
       "      <td>Format</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['file_format']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ent_1_model ent_1_handle ent_1_extra_handles ent_2_model ent_2_handle  \\\n",
       "0         DST   Identifier         ['Subject']        C2M2     local_id   \n",
       "1         DST         Type        ['Specimen']        C2M2   assay_type   \n",
       "2         DST    Ethnicity         ['Subject']        C2M2    ethnicity   \n",
       "3         DST         Race         ['Subject']        C2M2         race   \n",
       "4         DST       Format            ['File']        C2M2           id   \n",
       "\n",
       "  ent_2_extra_handles  \n",
       "0       ['biosample']  \n",
       "1       ['biosample']  \n",
       "2         ['subject']  \n",
       "3    ['subject_race']  \n",
       "4     ['file_format']  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c2m2_mapping_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Check original mapping spreadsheet and add any missing ents/mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>['Specimen']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>assay_type</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DST</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>['subject']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DST</td>\n",
       "      <td>Race</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>race</td>\n",
       "      <td>['subject_race']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DST</td>\n",
       "      <td>Format</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['file_format']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DST</td>\n",
       "      <td>Name</td>\n",
       "      <td>['Program']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DST</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>taxonomy_id</td>\n",
       "      <td>['subject_role_taxonomy']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>data_type</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DST</td>\n",
       "      <td>Name</td>\n",
       "      <td>['Study']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DST</td>\n",
       "      <td>Investigator Name</td>\n",
       "      <td>['Principal']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>contact_name</td>\n",
       "      <td>['dcc']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DST</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>['Specimen']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Study']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DST</td>\n",
       "      <td>Sex (Genotypic Sex)\\n</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>sex</td>\n",
       "      <td>['subject']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DST</td>\n",
       "      <td>symbol</td>\n",
       "      <td>['Gene']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['gene']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Project']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DST</td>\n",
       "      <td>Code</td>\n",
       "      <td>['Disease']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['disease']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>DST</td>\n",
       "      <td>Modality</td>\n",
       "      <td>['Imaging']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>mime_type</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>DST</td>\n",
       "      <td>age at specimen collection</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>age_at_sampling</td>\n",
       "      <td>['biosample_from_subject']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ent_1_model                ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0          DST                  Identifier         ['Subject']        C2M2   \n",
       "1          DST                        Type        ['Specimen']        C2M2   \n",
       "2          DST                   Ethnicity         ['Subject']        C2M2   \n",
       "3          DST                        Race         ['Subject']        C2M2   \n",
       "4          DST                      Format            ['File']        C2M2   \n",
       "5          DST                  Identifier            ['File']        C2M2   \n",
       "6          DST                        Name         ['Program']        C2M2   \n",
       "7          DST                       Taxon         ['Subject']        C2M2   \n",
       "8          DST                        Type            ['File']        C2M2   \n",
       "9          DST                        Name           ['Study']        C2M2   \n",
       "10         DST           Investigator Name       ['Principal']        C2M2   \n",
       "11         DST               Anatomic Site        ['Specimen']        C2M2   \n",
       "12         DST                  Identifier           ['Study']        C2M2   \n",
       "13         DST       Sex (Genotypic Sex)\\n         ['Subject']        C2M2   \n",
       "14         DST                      symbol            ['Gene']        C2M2   \n",
       "15         DST                  Identifier         ['Project']        C2M2   \n",
       "16         DST                        Code         ['Disease']        C2M2   \n",
       "17         DST                    Modality         ['Imaging']        C2M2   \n",
       "18         DST  age at specimen collection         ['Subject']        C2M2   \n",
       "\n",
       "       ent_2_handle         ent_2_extra_handles  \n",
       "0          local_id               ['biosample']  \n",
       "1        assay_type               ['biosample']  \n",
       "2         ethnicity                 ['subject']  \n",
       "3              race            ['subject_race']  \n",
       "4                id             ['file_format']  \n",
       "5          local_id                    ['file']  \n",
       "6              name                 ['project']  \n",
       "7       taxonomy_id   ['subject_role_taxonomy']  \n",
       "8         data_type                    ['file']  \n",
       "9              name                 ['project']  \n",
       "10     contact_name                     ['dcc']  \n",
       "11          anatomy               ['biosample']  \n",
       "12         local_id                 ['project']  \n",
       "13              sex                 ['subject']  \n",
       "14               id                    ['gene']  \n",
       "15         local_id                 ['project']  \n",
       "16               id                 ['disease']  \n",
       "17        mime_type                    ['file']  \n",
       "18  age_at_sampling  ['biosample_from_subject']  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c2m2_mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>File</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>mime_type</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DST</td>\n",
       "      <td>symbol</td>\n",
       "      <td>Gene</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>collection</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ent_1_model ent_1_handle ent_1_extra_handles ent_2_model ent_2_handle  \\\n",
       "0         DST         Type                File        C2M2    mime_type   \n",
       "1         DST       symbol                Gene        C2M2         name   \n",
       "2         DST   Identifier             Project        C2M2     local_id   \n",
       "\n",
       "  ent_2_extra_handles  \n",
       "0                file  \n",
       "1                gene  \n",
       "2          collection  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# additional lines to add (mapped in comments but not explicitly mapped in spreadsheet)\n",
    "ent_1_model = [\"DST\", \"DST\", \"DST\"]\n",
    "ent_1_handle = [\"Type\", \"symbol\", \"Identifier\"]\n",
    "ent_1_extra_handles = [\"File\", \"Gene\", \"Project\"]\n",
    "ent_2_model = [\"C2M2\", \"C2M2\", \"C2M2\"]\n",
    "ent_2_handle = [\"mime_type\", \"name\", \"local_id\"]\n",
    "ent_2_extra_handles = [\"file\", \"gene\", \"collection\"]\n",
    "\n",
    "extra_dict = {\n",
    "    \"ent_1_model\": ent_1_model,\n",
    "    \"ent_1_handle\": ent_1_handle,\n",
    "    \"ent_1_extra_handles\": ent_1_extra_handles,\n",
    "    \"ent_2_model\": ent_2_model,\n",
    "    \"ent_2_handle\": ent_2_handle,\n",
    "    \"ent_2_extra_handles\": ent_2_extra_handles\n",
    "}\n",
    "\n",
    "c2m2_extra = pd.DataFrame(extra_dict)\n",
    "\n",
    "c2m2_extra.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "c2m2_extra[\"ent_1_extra_handles\"] = [[l] for l in c2m2_extra[\"ent_1_extra_handles\"]] # listify extra handles\n",
    "c2m2_extra[\"ent_2_extra_handles\"] = [[l] for l in c2m2_extra[\"ent_2_extra_handles\"]] # listify extra handles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>[File]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>mime_type</td>\n",
       "      <td>[file]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DST</td>\n",
       "      <td>symbol</td>\n",
       "      <td>[Gene]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>[gene]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Project]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>[collection]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ent_1_model ent_1_handle ent_1_extra_handles ent_2_model ent_2_handle  \\\n",
       "0         DST         Type              [File]        C2M2    mime_type   \n",
       "1         DST       symbol              [Gene]        C2M2         name   \n",
       "2         DST   Identifier           [Project]        C2M2     local_id   \n",
       "\n",
       "  ent_2_extra_handles  \n",
       "0              [file]  \n",
       "1              [gene]  \n",
       "2        [collection]  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c2m2_extra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "c2m2_mapping_df_all = c2m2_mapping_df.append(c2m2_extra, ignore_index=True) # ignore index?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(c2m2_mapping_df_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>['Specimen']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>assay_type</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DST</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>['subject']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DST</td>\n",
       "      <td>Race</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>race</td>\n",
       "      <td>['subject_race']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DST</td>\n",
       "      <td>Format</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['file_format']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DST</td>\n",
       "      <td>Name</td>\n",
       "      <td>['Program']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DST</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>taxonomy_id</td>\n",
       "      <td>['subject_role_taxonomy']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>['File']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>data_type</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DST</td>\n",
       "      <td>Name</td>\n",
       "      <td>['Study']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DST</td>\n",
       "      <td>Investigator Name</td>\n",
       "      <td>['Principal']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>contact_name</td>\n",
       "      <td>['dcc']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DST</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>['Specimen']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>['biosample']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Study']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DST</td>\n",
       "      <td>Sex (Genotypic Sex)\\n</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>sex</td>\n",
       "      <td>['subject']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DST</td>\n",
       "      <td>symbol</td>\n",
       "      <td>['Gene']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['gene']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>['Project']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>['project']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DST</td>\n",
       "      <td>Code</td>\n",
       "      <td>['Disease']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>id</td>\n",
       "      <td>['disease']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>DST</td>\n",
       "      <td>Modality</td>\n",
       "      <td>['Imaging']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>mime_type</td>\n",
       "      <td>['file']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>DST</td>\n",
       "      <td>age at specimen collection</td>\n",
       "      <td>['Subject']</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>age_at_sampling</td>\n",
       "      <td>['biosample_from_subject']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DST</td>\n",
       "      <td>Type</td>\n",
       "      <td>[File]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>mime_type</td>\n",
       "      <td>[file]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>DST</td>\n",
       "      <td>symbol</td>\n",
       "      <td>[Gene]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>name</td>\n",
       "      <td>[gene]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DST</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Project]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>[collection]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ent_1_model                ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0          DST                  Identifier         ['Subject']        C2M2   \n",
       "1          DST                        Type        ['Specimen']        C2M2   \n",
       "2          DST                   Ethnicity         ['Subject']        C2M2   \n",
       "3          DST                        Race         ['Subject']        C2M2   \n",
       "4          DST                      Format            ['File']        C2M2   \n",
       "5          DST                  Identifier            ['File']        C2M2   \n",
       "6          DST                        Name         ['Program']        C2M2   \n",
       "7          DST                       Taxon         ['Subject']        C2M2   \n",
       "8          DST                        Type            ['File']        C2M2   \n",
       "9          DST                        Name           ['Study']        C2M2   \n",
       "10         DST           Investigator Name       ['Principal']        C2M2   \n",
       "11         DST               Anatomic Site        ['Specimen']        C2M2   \n",
       "12         DST                  Identifier           ['Study']        C2M2   \n",
       "13         DST       Sex (Genotypic Sex)\\n         ['Subject']        C2M2   \n",
       "14         DST                      symbol            ['Gene']        C2M2   \n",
       "15         DST                  Identifier         ['Project']        C2M2   \n",
       "16         DST                        Code         ['Disease']        C2M2   \n",
       "17         DST                    Modality         ['Imaging']        C2M2   \n",
       "18         DST  age at specimen collection         ['Subject']        C2M2   \n",
       "19         DST                        Type              [File]        C2M2   \n",
       "20         DST                      symbol              [Gene]        C2M2   \n",
       "21         DST                  Identifier           [Project]        C2M2   \n",
       "\n",
       "       ent_2_handle         ent_2_extra_handles  \n",
       "0          local_id               ['biosample']  \n",
       "1        assay_type               ['biosample']  \n",
       "2         ethnicity                 ['subject']  \n",
       "3              race            ['subject_race']  \n",
       "4                id             ['file_format']  \n",
       "5          local_id                    ['file']  \n",
       "6              name                 ['project']  \n",
       "7       taxonomy_id   ['subject_role_taxonomy']  \n",
       "8         data_type                    ['file']  \n",
       "9              name                 ['project']  \n",
       "10     contact_name                     ['dcc']  \n",
       "11          anatomy               ['biosample']  \n",
       "12         local_id                 ['project']  \n",
       "13              sex                 ['subject']  \n",
       "14               id                    ['gene']  \n",
       "15         local_id                 ['project']  \n",
       "16               id                 ['disease']  \n",
       "17        mime_type                    ['file']  \n",
       "18  age_at_sampling  ['biosample_from_subject']  \n",
       "19        mime_type                      [file]  \n",
       "20             name                      [gene]  \n",
       "21         local_id                [collection]  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c2m2_mapping_df_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Export clean csv with C2M2 mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "c2m2_mapping_df_all.to_csv(\"Mappings/c2m2_dss_mappings_clean.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run script to add mappings to MDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating new property node with properties: {handle: 'Identifier', model: 'DST', nanoid: 'atumqd', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'atumqd', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: '6f944Q', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'atumqd', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: '6f944Q', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'local_id', model: 'C2M2', nanoid: 'ghNETS'} to dst concept with properties: {nanoid: '6f944Q', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Type', model: 'DST', nanoid: '27bP76', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Specimen', model: 'DST', nanoid: 'kw8yZB', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Specimen', model: 'DST', nanoid: 'kw8yZB', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Type', model: 'DST', nanoid: '27bP76', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "One of the two entities not found in MDB, skipping linkage for now.\n",
      "Creating new property node with properties: {handle: 'Ethnicity', model: 'DST', nanoid: 'U7bwcR', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} to dst property with properties: {handle: 'Ethnicity', model: 'DST', nanoid: 'U7bwcR', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'YcBh1S', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Ethnicity', model: 'DST', nanoid: 'U7bwcR', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'YcBh1S', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'ethnicity', model: 'C2M2', nanoid: 'aS53di'} to dst concept with properties: {nanoid: 'YcBh1S', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Race', model: 'DST', nanoid: '3t0uqu', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} to dst property with properties: {handle: 'Race', model: 'DST', nanoid: '3t0uqu', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'eX912z', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Race', model: 'DST', nanoid: '3t0uqu', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'eX912z', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'race', model: 'C2M2', nanoid: 'Qrsf7S'} to dst concept with properties: {nanoid: 'eX912z', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Format', model: 'DST', nanoid: 'gdUzDM', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Format', model: 'DST', nanoid: 'gdUzDM', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: '7qYBga', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Format', model: 'DST', nanoid: 'gdUzDM', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: '7qYBga', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'id', model: 'C2M2', nanoid: 'DbmS3J'} to dst concept with properties: {nanoid: '7qYBga', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Identifier', model: 'DST', nanoid: 'sGGJgN', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB'} to dst property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'sGGJgN', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'RK6tth', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'sGGJgN', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'RK6tth', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'local_id', model: 'C2M2', nanoid: 'sBquyb'} to dst concept with properties: {nanoid: 'RK6tth', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Name', model: 'DST', nanoid: 'ZQbf8x', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Program', model: 'DST', nanoid: 'oMwkjS', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Program', model: 'DST', nanoid: 'oMwkjS', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Name', model: 'DST', nanoid: 'ZQbf8x', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: '2figiv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Name', model: 'DST', nanoid: 'ZQbf8x', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: '2figiv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'name', model: 'C2M2', nanoid: 'yi2WP1'} to dst concept with properties: {nanoid: '2figiv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Taxon', model: 'DST', nanoid: 'RuYx54', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} to dst property with properties: {handle: 'Taxon', model: 'DST', nanoid: 'RuYx54', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "One of the two entities not found in MDB, skipping linkage for now.\n",
      "Creating new property node with properties: {handle: 'Type', model: 'DST', nanoid: 'h5dwJF', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB'} to dst property with properties: {handle: 'Type', model: 'DST', nanoid: 'h5dwJF', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'i0x7E8', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Type', model: 'DST', nanoid: 'h5dwJF', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'i0x7E8', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'data_type', model: 'C2M2', nanoid: 'RbkPy9'} to dst concept with properties: {nanoid: 'i0x7E8', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Name', model: 'DST', nanoid: 'xNn4tz', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Study', model: 'DST', nanoid: 'i290aa', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Study', model: 'DST', nanoid: 'i290aa', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Name', model: 'DST', nanoid: 'xNn4tz', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Name', model: 'DST', nanoid: 'xNn4tz', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: '2figiv'}\n",
      "has_concept relationship already exists between src property with properties: {handle: 'name', model: 'C2M2', nanoid: 'yi2WP1'} to dst concept with properties: {nanoid: '2figiv'}\n",
      "Creating new property node with properties: {handle: 'Investigator Name', model: 'DST', nanoid: 'QZ1JXW', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Principal', model: 'DST', nanoid: 'a1VUar', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Principal', model: 'DST', nanoid: 'a1VUar', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Investigator Name', model: 'DST', nanoid: 'QZ1JXW', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'nPqeU6', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Investigator Name', model: 'DST', nanoid: 'QZ1JXW', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'nPqeU6', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'contact_name', model: 'C2M2', nanoid: '6GbXuT'} to dst concept with properties: {nanoid: 'nPqeU6', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Anatomic Site', model: 'DST', nanoid: 'w51AsC', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Specimen', model: 'DST', nanoid: 'kw8yZB'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Specimen', model: 'DST', nanoid: 'kw8yZB'} to dst property with properties: {handle: 'Anatomic Site', model: 'DST', nanoid: 'w51AsC', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'eEJ5rv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Anatomic Site', model: 'DST', nanoid: 'w51AsC', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'eEJ5rv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'anatomy', model: 'C2M2', nanoid: 'byu0ex'} to dst concept with properties: {nanoid: 'eEJ5rv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xotE4X', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Study', model: 'DST', nanoid: 'i290aa'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Study', model: 'DST', nanoid: 'i290aa'} to dst property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xotE4X', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'BBf9wC', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xotE4X', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'BBf9wC', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'local_id', model: 'C2M2', nanoid: 'HyBKbw'} to dst concept with properties: {nanoid: 'BBf9wC', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Sex (Genotypic Sex)\n",
      "', model: 'DST', nanoid: 'CfHTGb', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} to dst property with properties: {handle: 'Sex (Genotypic Sex)\n",
      "', model: 'DST', nanoid: 'CfHTGb', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: '1m6jN1', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Sex (Genotypic Sex)\n",
      "', model: 'DST', nanoid: 'CfHTGb', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: '1m6jN1', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'sex', model: 'C2M2', nanoid: 'w7QJtx'} to dst concept with properties: {nanoid: '1m6jN1', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'symbol', model: 'DST', nanoid: 'tAjqgM', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Gene', model: 'DST', nanoid: 'iyjkyr', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Gene', model: 'DST', nanoid: 'iyjkyr', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'symbol', model: 'DST', nanoid: 'tAjqgM', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'Mmfvg7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'symbol', model: 'DST', nanoid: 'tAjqgM', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'Mmfvg7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'id', model: 'C2M2', nanoid: 'UX3JgT'} to dst concept with properties: {nanoid: 'Mmfvg7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xUP6Si', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Project', model: 'DST', nanoid: 'd3rmnW', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Project', model: 'DST', nanoid: 'd3rmnW', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xUP6Si', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xUP6Si', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'BBf9wC'}\n",
      "has_concept relationship already exists between src property with properties: {handle: 'local_id', model: 'C2M2', nanoid: 'HyBKbw'} to dst concept with properties: {nanoid: 'BBf9wC'}\n",
      "Creating new property node with properties: {handle: 'Code', model: 'DST', nanoid: 'Ctv1X7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Disease', model: 'DST', nanoid: 'mgHhSC', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Disease', model: 'DST', nanoid: 'mgHhSC', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Code', model: 'DST', nanoid: 'Ctv1X7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'xV1em7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Code', model: 'DST', nanoid: 'Ctv1X7', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'xV1em7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'id', model: 'C2M2', nanoid: 'JPqFBV'} to dst concept with properties: {nanoid: 'xV1em7', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'Modality', model: 'DST', nanoid: 'Nq6EBv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new node node with properties: {handle: 'Imaging', model: 'DST', nanoid: 'Vtc5ZV', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_property relationship between src node with properties: {handle: 'Imaging', model: 'DST', nanoid: 'Vtc5ZV', _commit: '2022-11-30_c2m2-dss-map'} to dst property with properties: {handle: 'Modality', model: 'DST', nanoid: 'Nq6EBv', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new concept node with properties: {nanoid: 'iN5ejD', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'Modality', model: 'DST', nanoid: 'Nq6EBv', _commit: '2022-11-30_c2m2-dss-map'} to dst concept with properties: {nanoid: 'iN5ejD', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'mime_type', model: 'C2M2', nanoid: 'KjbkCB'} to dst concept with properties: {nanoid: 'iN5ejD', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Creating new property node with properties: {handle: 'age at specimen collection', model: 'DST', nanoid: 't95uUo', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "Node entity with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} already exists in the database.\n",
      "Adding has_property relationship between src node with properties: {handle: 'Subject', model: 'DST', nanoid: 'UA0baG'} to dst property with properties: {handle: 'age at specimen collection', model: 'DST', nanoid: 't95uUo', _commit: '2022-11-30_c2m2-dss-map'}\n",
      "One of the two entities not found in MDB, skipping linkage for now.\n",
      "Property entity with properties: {handle: 'Type', model: 'DST', nanoid: 'h5dwJF'} already exists in the database.\n",
      "Node entity with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB'} already exists in the database.\n",
      "has_property relationship already exists between src node with properties: {handle: 'File', model: 'DST', nanoid: 'vHtztB'} to dst property with properties: {handle: 'Type', model: 'DST', nanoid: 'h5dwJF'}\n",
      "has_concept relationship already exists between src property with properties: {handle: 'Type', model: 'DST', nanoid: 'h5dwJF'} to dst concept with properties: {nanoid: 'i0x7E8'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'mime_type', model: 'C2M2', nanoid: 'KjbkCB'} to dst concept with properties: {nanoid: 'i0x7E8'}\n",
      "Property entity with properties: {handle: 'symbol', model: 'DST', nanoid: 'tAjqgM'} already exists in the database.\n",
      "Node entity with properties: {handle: 'Gene', model: 'DST', nanoid: 'iyjkyr'} already exists in the database.\n",
      "has_property relationship already exists between src node with properties: {handle: 'Gene', model: 'DST', nanoid: 'iyjkyr'} to dst property with properties: {handle: 'symbol', model: 'DST', nanoid: 'tAjqgM'}\n",
      "has_concept relationship already exists between src property with properties: {handle: 'symbol', model: 'DST', nanoid: 'tAjqgM'} to dst concept with properties: {nanoid: 'Mmfvg7'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'name', model: 'C2M2', nanoid: 'WvNcfi'} to dst concept with properties: {nanoid: 'Mmfvg7'}\n",
      "Property entity with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xUP6Si'} already exists in the database.\n",
      "Node entity with properties: {handle: 'Project', model: 'DST', nanoid: 'd3rmnW'} already exists in the database.\n",
      "has_property relationship already exists between src node with properties: {handle: 'Project', model: 'DST', nanoid: 'd3rmnW'} to dst property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xUP6Si'}\n",
      "has_concept relationship already exists between src property with properties: {handle: 'Identifier', model: 'DST', nanoid: 'xUP6Si'} to dst concept with properties: {nanoid: 'BBf9wC'}\n",
      "Adding has_concept relationship between src property with properties: {handle: 'local_id', model: 'C2M2', nanoid: 'iS5W3R'} to dst concept with properties: {nanoid: 'BBf9wC'}\n"
     ]
    }
   ],
   "source": [
    "!python \"Script/link_synonym_ents_csv.py\" --csv_filepath=\"Mappings/c2m2_dss_mappings_clean.csv\" --mdb_uri=\"bolt://localhost:7687\" --mdb_user=\"neo4j\" --mdb_pass=\"blank1\" --entity_type=\"property\" --add_missing_ent_1=\"True\" --add_missing_ent_2=\"False\" --_commit=\"2022-11-30_c2m2-dss-map\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('bento_test_3_8')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ce2a05d834a6e1455d54a8cfaa6704fb454d813a6ed075fd3c8a17f8880dd21f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
